eip-0023.md

```diff
@@ -0,0 +1,530 @@
+- There is a group element in R4.
+
+The following (not enforced by the contract) must be done for a proper vote:
+- R5 contains the creation height of the current update box.
+- R6 contains the hash of the address of the new pool box.
+
+### Update Pool box
+
+| Index | Input | Output |
+|---|---|---|
+| 0 | Pool | New Pool |  
+| 1 | Update | Update |  
+| 2 | Ballot 1 | Ballot 1 |  
+| 3 | Ballot 2 | Ballot 2 |  
+| 4 | Ballot 3 | Ballot 3 |  
+| ... | ... | ... |
+
+This updates the pool box (i.e., transfers the poolNFT to a new address).
+
+The ballots used in inputs must satisfy the following as per the [smart contract](#ballot-contract):
+- There must be an output ballot box with the same public key in R4 and rest registers empty.
+
+The update box additionally ensures following as per the [smart contract](#update-contract):
+- Each ballot box has R4 containing the has of the new pool box script (address).
+- Each ballot box has R5 containing this box's creation height.
+- There is a copy of the update box in outputs with everything preserved but with larger creation height.
+- There are at least a threshold number of votes.
+- The registers and token in the pool box are preserved in the new pool box.
+
+We store the update box's creation height in the ballot's R5 to ensure that a vote is actually for the current update box.
+
+#### Update Rules
+
+In order to keep the [pool contract](#pool-contract) as compact as possible, many parameters are hard-wired in the [refresh contract](#refresh-contract). 
+Depending on what is to be updated, we may need to update other items as well. In particular, the following rules apply:
+
+| When updating | This must be updated | This need not be updated |
+|---|---|---|
+|Pool params (like epoch length) | Refresh-NFT <br /> Refresh box <br /> Reward tokens | Update-NFT <br /> Update box  <br /> Ballot tokens <br /> Ballot boxes <br /> Oracle tokens <br /> Oracle boxes (with new reward token) |
+|Refresh-NFT | Refresh box <br /> Reward tokens | Update-NFT <br /> Update box  <br /> Ballot tokens <br /> Ballot boxes <br /> Oracle tokens <br /> Oracle boxes (with new reward token) |
+|Update-NFT | Update box <br /> Ballot tokens <br /> Ballot boxes  | Refresh-NFT <br /> Refresh box <br /> Reward tokens <br /> Oracle tokens <br /> Oracle boxes |
+|Oracle tokens | Refresh-NFT <br /> Refresh box <br /> Reward tokens <br /> Oracle boxes <br /> | Update-NFT <br /> Update box <br /> Ballot tokens <br /> Ballot boxes |
+|Reward tokens | Refresh-NFT <br /> Refresh box | Update-NFT <br /> Update box <br /> Ballot tokens <br /> Ballot boxes <br /> Oracle tokens <br /> Oracle boxes |
+|Ballot tokens | Update-NFT <br /> Update box <br /> Ballot boxes | Refresh-NFT <br /> Refresh box <br /> Reward tokens <br /> Oracle tokens <br /> Oracle boxes |
+
+After an update, the older versions of the boxes will be left for garbage collection via storage-rent, and their tokens will become free for miners to take.
```

Comment:
I don't understand how to update a refresh box contract (to change epoch length, etc.). Should we mint a new Refresh-NFT and make a new refresh box, update pool box contract via voting ballots, mint new reward tokens, etc.? It looks like starting a new oracle pool should be easier.

Reply:
Yes new refresh NFT and new refresh box. Refer screenshot.
New oracle may be easier but not always possible because oracleNFT is hardwired in dApps

![Screenshot from 2022-05-11 21-51-14](https://user-images.githubusercontent.com/23208922/167899248-77cc5b12-0017-4cc8-8bad-638b0d4995d1.png)

Reply:
Sure, I saw the table. I just found it's incomplete for the changes in the refresh contract.
Let's see what we need to do in order to change the epoch length:
- mint new refresh nft token;
- update pool contract/box (refresh nft is hard-coded the pool contract);
- mint new reward tokens;
- **force oracle boxes to use the new reward token; How? I did not find a way yet.**
- make a new refresh box;

As you can see there is one show-stopper (oracle boxes) and all these steps need to be done while oracles are running and can spend the old refresh box at any time.

Reply:
Delegating the switch to a new reward token for the oracles is certainly the way. I just wanted to find a way to keep the whole "change epoch length" operation local, i.e. run by one party. Which in hindsight kind of contradicts the decentralized notion of the oracle pool. 

So, let's iterate on the workflow needed to change the epoch length:
1. Mint the new refresh nft;
2. Run voting via ballots and update the pool contract/box;
3. Mint new reward tokens;
4. Send one new reward token to each oracle (to be put in the new oracle box in the next step);
5. Coordinate with **every** oracle to run a command to create a new oracle box and transfer oracle nft for each oracle;
6. Make a new refresh box;

I find this workflow quite complex. Also, issuing new reward tokens bothers me. I assume there will be some infrastructure built around reward tokens to make a profit for oracles. It should be updated as well in order to transition to the new reward token, right?  

Can we find a way to avoid issuing new reward tokens and updating oracle boxes in this case (updating pool parameters)?

Reply:
Have updated contracts. Please review again